### ISSUE TITLE
[FR] expand: more flexible size inference 

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#18901]
## 🚀 Feature
`expand` could just match the `-1`-size dimensions by order, rather than by exact dimension index. 

E.g., make this work:
```
In [15]: x = torch.empty(2, 3, 4)

In [16]: x.expand(-1, 10, -1, -1)  # give me a [2, 10, 3, 4] tensor
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-16-d02944043762> in <module>
----> 1 x.expand(-1, 10, -1, -1)  # give me a [2, 10, 3, 4] tensor

RuntimeError: The expanded size of the tensor (10) must match the existing size (2) at non-singleton dime
nsion 1.  Target sizes: [-1, 10, -1, -1].  Tensor sizes: [2, 3, 4]
```

## Motivation
It makes expanding a dimension much more flexible. Otherwise I would need to chain an `unsqueeze` with an `expand`.

**Comment by zou3519** [comment]
Would we require that there needs to be exactly `x.dim()` -1's? For example, what should happen with the following?
```
x = torch.empty(2, 2, 3)
x.expand(-1, 3, -1)
```

**Comment by SsnL** [comment]
@zou3519 The old rule works by (for `-1` sizes) simply copying dim sizes of the old tensor to the expanded size. I.e., it only applies when (1) number of `-1` < `x.dim()`, or (2) non-`-1` sizes are followed by only all `x.dim()` consecutive `-1`s. (1) doesn't conflict with the rule I proposed. Behavior of (2) is the same for original rule and the rule I proposed. 

So this isn't conflicting with the existing behavior, but it could be tricky to understand.
